kratos/python_scripts/restart_utility.py

```diff
@@ -170,6 +176,50 @@ def CreateOutputFolder(self):
                 os.makedirs(folder_path)
             self.model_part.GetCommunicator().GetDataCommunicator().Barrier()
 
+    def GetRestartFiles(self):
+        """
+        Return a list of all restart files as os.DirEntry objects sorted in ascending order.
+        """
+        restart_files = []
+        if os.path.isdir(self.__GetFolderPathSave()):
+            with os.scandir(path=self.__GetFolderPathSave()) as contents:                           # <-- list all restart files
+                for entry in contents:
+                    if self.__IsRestartFile(entry):
+                        restart_files.append(entry)
+
+            modified_key = lambda entry: entry.stat().st_mtime_ns                                   # <-- sort files
+            restart_files.sort( key=modified_key )                                                  # - by time of last modification
+
+            def label_key(dir_entry):
+                key = self.__ExtractFileLabel(dir_entry.name)                                       # - by their labels
+                if key is None:                                                                     # (time of modification might be ambiguous)
+                    key = 0
+                return key
+            restart_files.sort( key=label_key )
+        return restart_files
+
+    def ClearObsoleteRestartFiles(self):
+        """
+        Collect all restart files from the current restart directory, sort them by date(time) modified
+        and delete the oldest ones such that only number_of_stored_files remain.
+        Note: a secondary sorting is performed based on the labels of the file names to resolve equalities.
+        """
+        if self.number_of_stored_files > -1:
+            restart_files = self.GetRestartFiles()
+            number_of_obsolete_files = len(restart_files) - self.number_of_stored_files
```

Comment:
I am not sure this will work as intended with MPI: there one will have `number_of_cpu*self.number_of_restart_files`. So maybe checking all the files in the directory is not the easiest way to go. Using a buffer-type storage will lead to the problem when actually restarting, so that is not the best either.

For a different process I write out a test file which is read-in at restart. Could this be an easy and straightforward option for logging the already written out file names? This would shift this part from the need of scanning a directory to an read-write operation. Could be done with the usage of [https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/python_scripts/time_based_ascii_file_writer_utility.py](url) -> Storing time stamp and restart file label.

Sample labeling for MPI:

```
...
FluidModelPart_95_150.0.rest  
FluidModelPart_96_150.0.rest
FluidModelPart_97_150.0.rest
FluidModelPart_98_150.0.rest
FluidModelPart_99_150.0.rest
...
```
==>> `<model_part_name>_<cpu_number>_<time_stamp>.rest`



@philbucher ?

Reply:
I can't find the file you're referring to.

In that case, we'll need a unit test for parallel executions too. Can you please show me how to configure to run with MPI?

Just to be clear: when running with MPI, are all restart files generated at a specific step needed to resume the program? The option name ```number_of_restart_files``` would be a bit misleading then, so I'd change it to ```number_of_restart_file_sets``` if that's okay.

I could go for the solution with an extra log file you mentioned, or alternatively, I could modify the current setup to use a buffer that is initialized with the current folder scanning process at the start of the run. Let me know which one you'd prefer.

Reply:
The naming of the restart files is:
`<local_model_part> = <model_part>_<processor_number>`

Because the save and load model part is actually executed on each processor (MPI) only the time stamp and format is added.

I formulated my comment previously a less precise way.

Solution with a written log-file: each processor (for MPI) will write one, so I am - after some deliberating - not too much of a fan. Still, if you want to have a got, start from here [time_based_ascii_file_writer_utility.py](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/python_scripts/time_based_ascii_file_writer_utility.py) (if the hyperlink does not work, just copy the full path, seeing how the [point_ouput_process.py]() uses it at restart. One could write out the time, step, label.

Solution with looking in the folder, checking for files, filtering and validating names: I would only do it when the restart is loaded (so checking the flag is restart), otherwise I think it is too lengthy.

Solution with some kind of buffer in the class which is preallocated and refreshed: perhaps this would be the preferred way during run-time. The problem here could be that, potentially, one could have a restart for many steps, which means a large preallocated python list (could become inefficient). So a warning could be good to have if either all steps or many (more than 1000?) files are needed.